### PR TITLE
curvefs client : fix bug of getleader always fails causes stack overflow

### LINKED ISSUE #1067: curve-fuse getleader always fails causes stack overflow
**Describe the bug (描述bug)**

curve-fuse getleader always fails causes stack overflow

**To Reproduce (复现方法)**

stop all metaserver, and write some things

**Expected behavior (期望行为)**

curve-fuse stack overflow

**Versions (各种版本)**
OS:
Compiler:
curve-mds:
curve-chunkserver:
curve-snapshotcloneserver:
curve-sdk:
nebd:
curve-nbd:

**Additional context/screenshots (更多上下文/截图)**

image

### PR DESCRIPTION
**Pull request by xu-chaojie**
### What problem does this PR solve?

Issue Number: close #1067 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by xu-chaojie** [comment]
recheck

File: curvefs/src/client/rpcclient/metaserver_client.cpp [hunk 1]
```diff
@@ -586,8 +587,16 @@ void MetaServerClientImpl::UpdateInodeAsync(const Inode &inode,
         MetaServerOpType::UpdateInode, task, inode.fsid(), inode.inodeid());
     auto excutor = std::make_shared<UpdateInodeExcutor>(opt_,
         metaCache_, channelManager_, taskCtx);
-    TaskExecutorDone *taskDone = new TaskExecutorDone(excutor, done);
-    excutor->DoAsyncRPCTask(taskDone);
+    TaskExecutorDone *taskDone = new TaskExecutorDone(
+        excutor, done);
+    done_guard.release();
```
Comment (wu-hanqing) [inline curvefs/src/client/rpcclient/metaserver_client.cpp:592]: it seems this `done_guard` is useless

Comment (xu-chaojie) [reply curvefs/src/client/rpcclient/metaserver_client.cpp:592]: done

File: curvefs/src/client/rpcclient/metaserver_client.cpp [hunk 2]
```diff
@@ -741,8 +751,16 @@ void MetaServerClientImpl::GetOrModifyS3ChunkInfoAsync(
         MetaServerOpType::GetOrModifyS3ChunkInfo, task, fsId, inodeId);
     auto excutor = std::make_shared<GetOrModifyS3ChunkInfoExcutor>(opt_,
         metaCache_, channelManager_, taskCtx);
-    TaskExecutorDone *taskDone = new TaskExecutorDone(excutor, done);
-    excutor->DoAsyncRPCTask(taskDone);
+    TaskExecutorDone *taskDone = new TaskExecutorDone(
+        excutor, done);
+    done_guard.release();
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply curvefs/src/client/rpcclient/metaserver_client.cpp:756]: done

File: curvefs/src/client/rpcclient/task_excutor.cpp [hunk 3]
```diff
@@ -79,37 +79,38 @@ int TaskExecutor::DoRPCTask() {
...
-    }
+        auto channel = channelManager_->GetOrCreateChannel(
+            task_->target.metaServerID, task_->target.endPoint);
+        if (!channel) {
+            LOG(WARNING) << "GetOrCreateChannel fail for "
+                         << task_->TaskContextStr() << ", sleep and retry";
+            bthread_usleep(opt_.retryIntervalUS);
+            continue;
+        }
+        ExcuteTask(channel.get(), done);
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply curvefs/src/client/rpcclient/task_excutor.cpp:109]: done

File: curvefs/src/client/rpcclient/metaserver_client.cpp [hunk 4]
```diff
@@ -741,8 +751,16 @@ void MetaServerClientImpl::GetOrModifyS3ChunkInfoAsync(
...
+    brpc::ClosureGuard taskDone_guard(taskDone);
+    int ret = excutor->DoAsyncRPCTask(taskDone);
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply curvefs/src/client/rpcclient/metaserver_client.cpp:758]: if getleader failed, which means the cluster is abnomal, it is reasonable to stuck  all the async task here.

File: curvefs/src/client/rpcclient/task_excutor.cpp [hunk 5]
```diff
@@ -79,37 +79,38 @@ int TaskExecutor::DoRPCTask() {
...
+        break;
+    } while (true);
 
-    ExcuteTask(channel.get(), done);
-    done_guard.release();
-    return;
+    return retCode;
```
[the remark this answers is not in the document]
Comment (xu-chaojie) [reply curvefs/src/client/rpcclient/task_excutor.cpp:113]: done

**Comment by xu-chaojie** [comment]
recheck

**Comment by xu-chaojie** [comment]
recheck

**Review by wu-hanqing — APPROVED** [review]

File: curvefs/src/client/rpcclient/task_excutor.cpp [hunk 6]
```diff
@@ -303,7 +304,11 @@ void TaskExecutorDone::Run() {
     needRetry = excutor_->OnReturn(code_);
     if (needRetry) {
         excutor_->PreProcessBeforeRetry(code_);
-        excutor_->DoAsyncRPCTask(this);
+        code_ = excutor_->DoAsyncRPCTask(this);
+        if (code_ < 0) {
```
Comment (ilixiaocui) [inline curvefs/src/client/rpcclient/task_excutor.cpp:308]: AsyncTask always return MetaStatusCode::OK
https://github.com/opencurve/curve/blob/731b85c5eed0d632d4f429d72b12d6fdab8882e2/curvefs/src/client/rpcclient/metaserver_client.cpp#L582

https://github.com/opencurve/curve/blob/731b85c5eed0d632d4f429d72b12d6fdab8882e2/curvefs/src/client/rpcclient/metaserver_client.cpp#L737

Comment (xu-chaojie) [reply curvefs/src/client/rpcclient/task_excutor.cpp:308]: AsyncTask will return -1， when retryTimes runing out

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**